external-crates/move/crates/move-compiler/src/parser/ast.rs

```diff
@@ -650,6 +711,253 @@ impl LeadingNameAccess_ {
+        tyargs: Option<Spanned<Vec<Type>>>,
+        is_macro: Option<Loc>,
+    ) -> Vec<Diagnostic> {
+        let mut diags: Vec<Diagnostic> = vec![];
+
+        let mut final_tyargs = tyargs;
+
+        if let (Some(prev_loc), Some(sp!(new_loc, _))) = (self.tyargs_loc(), &final_tyargs) {
+            let mut diag = diag!(
+                Syntax::InvalidName,
+                (
+                    *new_loc,
+                    "Paths cannot include type arguments more than once"
+                ),
+                (prev_loc, "Previous type arguments appeared here")
+            );
+            diag.add_note("Type arguments should only appear on module members");
+            diags.push(diag);
+            // If we already had tyargs, remove these.
+            final_tyargs = None;
+        }
+
+        if let Some(prev_loc) = self.is_macro() {
+            let diag = diag!(
+                Syntax::InvalidName,
+                (
+                    name.loc,
+                    "A macro call cannot have name access entries after it"
+                ),
+                (*prev_loc, "Macro invocation given here")
+            );
+            diags.push(diag);
+            // If this is a macro, remove previous `!` usages.
+            self.root.is_macro = None;
+            for entry in self.entries.iter_mut() {
+                entry.is_macro = None;
+            }
+        }
+
+        if self.len() > 3 {
+            let diag = diag!(
+                Syntax::InvalidName,
+                (name.loc, "Paths cannot have length greater than four")
+            );
+            diags.push(diag);
+        } else {
+            let path_entry = PathEntry {
+                name,
+                tyargs: final_tyargs,
+                is_macro,
+            };
+            self.entries.push(path_entry);
+        }
+        diags
+    }
+
+    fn has_tyargs(&self) -> bool {
+        self.root.tyargs.is_some() || self.entries.iter().any(|entry| entry.tyargs.is_some())
+    }
+
+    fn has_tyargs_last(&self) -> bool {
+        if !self.has_tyargs() {
+            // Tyargs are last vacuously
+            true
+        } else if let Some(last) = self.entries.last() {
+            last.tyargs.is_some()
+        } else {
+            self.entries.is_empty() && self.root.tyargs.is_some()
+        }
+    }
+
+    fn tyargs_loc(&self) -> Option<Loc> {
+        if let Some(sp!(loc, _)) = self.root.tyargs {
+            return Some(loc);
+        }
+        for entry in self.entries.iter() {
+            if let Some(sp!(loc, _)) = entry.tyargs {
+                return Some(loc);
+            }
+        }
+        None
+    }
+
+    fn len(&self) -> usize {
+        1 + self.entries.len()
+    }
+}
+
+macro_rules! forward_name_access {
+    ($self:ident.$call:ident($($args:ident),*)) => {
+        match $self {
+            NameAccessChain_::Single(entry) => entry.$call($($args),*),
+            NameAccessChain_::Path(entry) => entry.$call($($args),*),
```

Comment:
It might not be called enum_dispatch... but I know what it is ;)

Reply:
Honestly I probably prefer this lol

Reply:
Yeah, I think I like this more, too. The overall thing is clear without too much "magic."